File: packages/ios/OpenCode/API/EventStream.swift
import Foundation

/// Minimal Server-Sent Events reader over URLSession's async byte stream.
///
/// Yields the raw JSON payload of each `data:` frame as `Data` (which is
/// `Sendable`); the caller decodes it into a `ServerEvent` on its own actor.
/// Cancelling the consuming task tears down the underlying URLSession request.
struct EventStream {
    let url: URL
    let authHeader: String?

    func frames() -> AsyncThrowingStream<Data, Error> {
        AsyncThrowingStream { continuation in
            let task = Task {
                do {
                    var request = URLRequest(url: url)
                    request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
                    // Long-lived stream: keep the request from timing out while idle.
                    request.timeoutInterval = 86_400
                    if let authHeader {
                        request.setValue(authHeader, forHTTPHeaderField: "Authorization")
                    }

                    let (bytes, response) = try await URLSession.shared.bytes(for: request)
                    guard let http = response as? HTTPURLResponse,
                          (200...299).contains(http.statusCode) else {
                        throw ClientError.http((response as? HTTPURLResponse)?.statusCode ?? 0)
                    }

                    // Parse the SSE stream byte-by-byte rather than via
                    // `bytes.lines`: that async sequence withholds the blank
                    // line that terminates a frame until more input arrives, so
                    // events would only surface one event late (or never, while
                    // the model is idle). Splitting on `\n` ourselves dispatches
                    // each frame the moment its terminating blank line lands.
                    var line = Data()         // bytes of the current line (no \n)
                    var frame = Data()         // accumulated `data:` payload(s)

                    func flushLine() {
                        defer { line.removeAll(keepingCapacity: true) }
                        // Tolerate CRLF.
                        if line.last == 0x0D { line.removeLast() }
                        if line.isEmpty {
                            // Blank line terminates an event.
                            if !frame.isEmpty {
                                continuation.yield(frame)
                                frame.removeAll(keepingCapacity: true)
                            }
                            return
                        }
                        guard line.starts(with: Data("data:".utf8)) else { return } // ignore :, event:, id:, retry:
                        var payload = line.dropFirst("data:".count)
                        while payload.first == 0x20 { payload = payload.dropFirst() } // trim leading spaces
                        if !frame.isEmpty { frame.append(0x0A) } // join multi-line data with \n
                        frame.append(contentsOf: payload)
                    }

                    for try await byte in bytes {
                        try Task.checkCancellation()
                        if byte == 0x0A { // \n
                            flushLine()
                        } else {
                            line.append(byte)
                        }
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }
}

File: packages/android/app/src/main/java/studio/eugenezakharov/opencode/api/EventStream.kt
package studio.eugenezakharov.opencode.api

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.Dispatchers
import okhttp3.OkHttpClient
import okhttp3.Request
import java.util.concurrent.TimeUnit

/**
 * Minimal Server-Sent Events reader over OkHttp's streaming response body.
 *
 * Emits the raw JSON payload of each `data:` frame as a [String]; the caller
 * decodes it into a [ServerEvent].
 *
 * The stream is parsed byte/line-by-line and a frame is dispatched **the moment
 * its terminating blank line arrives** — not via a buffered line reader that
 * withholds the blank line until more input lands (the exact bug that bit the
 * iOS client: events would surface one event late, or never while idle).
 */
class EventStream(
    private val url: String,
    private val authHeader: String?,
    /** In relay mode, the per-tunnel token the relay requires (`X-Tunnel-Token`); null in direct mode. */
    private val tunnelToken: String? = null,
) {
    private val client = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        // Long-lived stream: no read timeout so it never trips while idle.
        .readTimeout(0, TimeUnit.MILLISECONDS)
        .retryOnConnectionFailure(true)
        .build()

    fun frames(): Flow<String> = flow {
        val requestBuilder = Request.Builder()
            .url(url)
            .header("Accept", "text/event-stream")
        authHeader?.let { requestBuilder.header("Authorization", it) }
        tunnelToken?.let { requestBuilder.header("X-Tunnel-Token", it) }

        client.newCall(requestBuilder.build()).execute().use { response ->
            if (!response.isSuccessful) throw ClientError.Http(response.code)
            val source = response.body?.source() ?: throw ClientError.Http(0)

            val frame = StringBuilder()
            // BufferedSource.readUtf8Line() returns one line per call as soon as
            // the line terminator (\n, tolerating \r\n) arrives, and returns an
            // empty string for a blank line — so a frame is dispatched exactly
            // when its terminating blank line lands.
            while (true) {
                val line = source.readUtf8Line() ?: break // stream closed
                if (line.isEmpty()) {
                    // Blank line terminates an event.
                    if (frame.isNotEmpty()) {
                        emit(frame.toString())
                        frame.setLength(0)
                    }
                    continue
                }
                // Ignore comments (`:`), `event:`, `id:`, `retry:` — only `data:` carries the payload.
                if (!line.startsWith("data:")) continue
                var payload = line.substring("data:".length)
                if (payload.startsWith(" ")) payload = payload.substring(1)
                if (frame.isNotEmpty()) frame.append('\n') // join multi-line data with \n
                frame.append(payload)
            }
        }
    }.flowOn(Dispatchers.IO)
}
